src/simulators/ch/ch_state.hpp

```diff
@@ -0,0 +1,787 @@
+                                      OutputData &data,
+                                      RngEngine &rng);
+  // Applies a sypported Gate operation to the state class.
+  // If the input is not in allowed_gates an exeption will be raised.
+  // TODO: Investigate OMP synchronisation over stattes to remove these different versions
+  // One option would be tasks, but the memory overhead isn't clear
+  void apply_gate(const Operations::Op &op, RngEngine &rng);
+  void apply_gate(const Operations::Op &op, RngEngine &rng, uint_t rank);
+
+  // Measure qubits and return a list of outcomes [q0, q1, ...]
+  // If a state subclass supports this function then "measure" 
+  // should be contained in the set returned by the 'allowed_ops'
+  // method.
+  void apply_measure(const reg_t &qubits,
+                     const reg_t &cmemory,
+                     const reg_t &cregister,
+                     RngEngine &rng);
+
+  // Reset the specified qubits to the |0> state by measuring the
+  // projectors Id+Z_{i} for each qubit i
+  void apply_reset(const reg_t &qubits, AER::RngEngine &rng);
+
+  //Take a snapshot of the simulation state
+  //TODO: Improve the CHSimulator::to_json method.
+  void apply_snapshot(const Operations::Op &op, OutputData &data, RngEngine &rng);
+  //Convert a decomposition to a state-vector
+  void statevector_snapshot(const Operations::Op &op, OutputData &data, RngEngine &rng);
+  //Compute probabilities from a stabilizer rank decomposition
+  //TODO: Check ordering/output format...
+  void probabilities_snapshot(const Operations::Op &op, OutputData &data, RngEngine &rng);
+
+  const static stringmap_t<Gates> gateset_;
+  const static stringmap_t<Snapshots> snapshotset_;
+
+  //-----------------------------------------------------------------------
+  //Parameters and methods specific to the Stabilizer Rank Decomposition
+  //-----------------------------------------------------------------------
+
+  //Allowed error in the stabilizer rank decomposition.
+  //The required number of states scales as \delta^{-2}
+  //for allowed error \delta
+  double approximation_error = 0.05;
+
+  uint_t norm_estimation_samples = 100;
+  // How long the metropolis algorithm runs before
+  // we consider it to be well mixed and sample form the
+  // output distribution
+  uint_t metropolis_mixing_steps = 7000;
+
+  //Minimum number of states before we try to parallelise
+  uint_t omp_threshold_rank = 100;
+
+  double snapshot_chop_threshold = 1e-10;
+
+  double probabilities_snapshot_samples = 3000.;
+
+  // Compute the required stabilizer rank of the circuit
+  uint_t compute_chi(const std::vector<Operations::Op> &ops);
+  // Add the given operation to the extent
+  inline void compute_extent(const Operations::Op &op, double &xi);
+
+  //Compute the required chi, and count the number of three qubit gates
+  virtual std::pair<uint_t, uint_t> decomposition_parameters(const std::vector<Operations::Op> &ops);
```

Comment:
I'd remove `virtual` keyword here, if we are not planning to subclass this class.

Reply:
I think this can be `const` as well